api/v1beta1/bucket_types.go

```diff
@@ -106,22 +107,22 @@ const (
 	BucketOperationFailedReason string = "BucketOperationFailed"
 )
 
-// BucketProgressing resets the conditions of the Bucket to meta.Condition of
+// BucketProgressing resets the conditions of the Bucket to metav1.Condition of
 // type meta.ReadyCondition with status 'Unknown' and meta.ProgressingReason
 // reason and message. It returns the modified Bucket.
 func BucketProgressing(bucket Bucket) Bucket {
 	bucket.Status.ObservedGeneration = bucket.Generation
 	bucket.Status.URL = ""
-	bucket.Status.Conditions = []meta.Condition{}
-	SetBucketCondition(&bucket, meta.ReadyCondition, corev1.ConditionUnknown, meta.ProgressingReason, "reconciliation in progress")
+	bucket.Status.Conditions = []metav1.Condition{}
+	SetBucketCondition(&bucket, meta.ReadyCondition, metav1.ConditionUnknown, meta.ProgressingReason, "reconciliation in progress")
 	return bucket
 }
 
 // SetBucketCondition sets the given condition with the given status, reason and
 // message on the Bucket.
-func SetBucketCondition(bucket *Bucket, conditionType string, status corev1.ConditionStatus, reason, message string) {
+func SetBucketCondition(bucket *Bucket, conditionType string, status metav1.ConditionStatus, reason, message string) {
```

Comment:
Can we remove this method (and the corresponding methods for the other types) and just use `metav1.SetStatusCondition()` instead?

Reply:
Or perhaps use `metav1.SetStatusCondition()` within these methods. Or if possible add a convenience method to `pkg/runtime` which has the same signature as these methods, except with a loosely typed first argument and gets the metadata generation and status conditions, and then calls `metav1.SetStatusCondition()`.

Reply:
Makes sense to me. Will follow-up with the changes.

Reply:
I think we can also remove `FilterOutCondition` from `fluxcd/pkg` now that we have `meta.RemoveStatusCondition()`?

Reply:
> I think we can also remove `FilterOutCondition` from `fluxcd/pkg` now that we have `meta.RemoveStatusCondition()`?

If that was its intended purpose then sure. From the implementation (and name) it was not that clear, to be honest.

Reply:
I missed this during my review. But they behave indeed the same, except that the one in `fluxcd/pkg` returns a new slice and the one from the Kubernetes upstream _modifies_ the given slice. The latter is more Kubernetes idiomatic, and has my preference (we should work with pointers as much as we can actually).

Reply:
Here's the removal of FilterOutCondition: https://github.com/fluxcd/pkg/pull/47

Reply:
> Or perhaps use `metav1.SetStatusCondition()` within these methods. Or if possible add a convenience method to `pkg/runtime` which has the same signature as these methods, except with a loosely typed first argument and gets the metadata generation and status conditions, and then calls `metav1.SetStatusCondition()`.

I suppose having a dedicated convenience method in `pkg` would be nice if we actually do want to handle updating the `observedGeneration` of the conditions as well during.

Reply:
On a side note, looks like updating the `ObservedGeneration` in the `SetStatusCondition` func was [added quite recently to v1.20](https://github.com/kubernetes/kubernetes/commit/5712e33abcea86e7bf699f40a3a838afc1b7c278).

Reply:
> Or perhaps use `metav1.SetStatusCondition()` within these methods. Or if possible add a convenience method to `pkg/runtime` which has the same signature as these methods, except with a loosely typed first argument and gets the metadata generation and status conditions, and then calls `metav1.SetStatusCondition()`.

Implemented here: https://github.com/fluxcd/pkg/pull/48